shell/platform/common/text_input_model.cc

```diff
@@ -69,15 +69,22 @@ void TextInputModel::BeginComposing() {
   composing_range_ = TextRange(selection_.start());
 }
 
-void TextInputModel::UpdateComposingText(const std::u16string& text) {
+void TextInputModel::UpdateComposingText(const std::u16string& text,
+                                         const TextRange& selection) {
   // Preserve selection if we get a no-op update to the composing region.
   if (text.length() == 0 && composing_range_.collapsed()) {
     return;
   }
-  DeleteSelected();
-  text_.replace(composing_range_.start(), composing_range_.length(), text);
+  const TextRange& rangeToDelete =
+      composing_range_.collapsed() ? selection_ : composing_range_;
+  text_.replace(rangeToDelete.start(), rangeToDelete.length(), text);
   composing_range_.set_end(composing_range_.start() + text.length());
-  selection_ = TextRange(composing_range_.end());
+  selection_ = TextRange(selection.start() + composing_range_.start(),
+                         selection.extent() + composing_range_.start());
+}
```

Comment:
It might be useful to add very short one-line comments here to help understanding for developers who don't routinely work with CJK languages that use composing, specifically covering the relationship/interaction between composing range and selection.

Reply:
This is a public method so the exact behavior should already be documented in the header file?

Reply:
This is what I get for reviewing .cc before .h files. The doc comments looked great.